File: packages/sdkwork-utils-java/src/main/java/com/sdkwork/utils/NumberUtils.java
package com.sdkwork.utils;

public final class NumberUtils {
    private NumberUtils() {
    }

    public static double clamp(double value, double min, double max) {
        return Math.min(max, Math.max(min, value));
    }

    public static double round(double value, int decimals) {
        double factor = Math.pow(10, decimals);
        return Math.round(value * factor) / factor;
    }

    public static String formatNumber(double value, int decimals) {
        return String.format("%." + decimals + "f", value);
    }

    public static Double parseNumber(String value) {
        try {
            return Double.parseDouble(value.trim());
        } catch (NumberFormatException ex) {
            return null;
        }
    }

    public static boolean isInteger(double value) {
        return Double.isFinite(value) && value == Math.rint(value);
    }

    public static Long parseInt(String value) {
        try {
            return Long.parseLong(value.trim());
        } catch (NumberFormatException ex) {
            return null;
        }
    }

    public static String percentFormat(double value, int decimals) {
        return formatNumber(value * 100.0, decimals) + "%";
    }

    public static boolean inRange(double value, double min, double max) {
        return value >= min && value <= max;
    }

    public static double abs(double value) {
        return Math.abs(value);
    }
}

File: packages/sdkwork-utils-kotlin/src/main/kotlin/com/sdkwork/utils/NumberUtils.kt
package com.sdkwork.utils

object NumberUtils {
    fun clamp(value: Double, min: Double, max: Double): Double {
        return minOf(max, maxOf(min, value))
    }

    fun round(value: Double, decimals: Int): Double {
        val factor = Math.pow(10.0, decimals.toDouble())
        return Math.round(value * factor) / factor
    }

    fun formatNumber(value: Double, decimals: Int): String {
        return String.format("%.${decimals}f", value)
    }

    fun parseNumber(value: String): Double? {
        return try {
            value.trim().toDouble()
        } catch (_: NumberFormatException) {
            null
        }
    }

    fun isInteger(value: Double): Boolean {
        return value.isFinite() && value == Math.rint(value)
    }

    fun parseInt(value: String): Long? {
        return try {
            value.trim().toLong()
        } catch (_: NumberFormatException) {
            null
        }
    }

    fun percentFormat(value: Double, decimals: Int): String {
        return "${formatNumber(value * 100.0, decimals)}%"
    }

    fun inRange(value: Double, min: Double, max: Double): Boolean {
        return value in min..max
    }

    fun abs(value: Double): Double = kotlin.math.abs(value)
}
